sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java

```diff
@@ -75,191 +83,230 @@ public static Row nullRow(RowType rowType) {
-   * if type doesn't match.
+   * Get a {@link TypeName#INT16} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
-  public Short getShort(int idx) {
+  public Short getInt16(int idx) {
+    checkState(
+        TypeName.INT16.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Integer} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#INT32} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
-  public Integer getInteger(int idx) {
+  public Integer getInt32(int idx) {
+    checkState(
+        TypeName.INT32.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Float} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#FLOAT} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
   public Float getFloat(int idx) {
+    checkState(
+        TypeName.FLOAT.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Double} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#DOUBLE} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
   public Double getDouble(int idx) {
+    checkState(
+        TypeName.DOUBLE.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Long} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#INT64} value by field index, {@link ClassCastException} is thrown
+   * if schema doesn't match.
    */
-  public Long getLong(int idx) {
+  public Long getInt64(int idx) {
+    checkState(
+        TypeName.INT64.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
    * Get a {@link String} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * if schema doesn't match.
    */
   public String getString(int idx) {
+    checkState(
+        TypeName.STRING.equals(getSchema().getField(idx).getTypeDescriptor().getType()));
     return getValue(idx);
   }
 
   /**
-   * Get a {@link Date} value by field index, {@link ClassCastException} is thrown
-   * if type doesn't match.
+   * Get a {@link TypeName#DATETIME} value by field index, {@link IllegalStateException} is thrown
+   * if schema doesn't match.
+   */
+  public ReadableDateTime getDateTime(int idx) {
+    ReadableInstant instant = getValue(idx);
+    return new DateTime(instant).withZone(instant.getZone());
```

Comment:
Why can't we either return ReadableInstant or store DateTime or ReadableDateTime in the first place?

Reply:
essentially ReadableDateTime is just extra convenience methods on top of ReadableInstant. I think it makes sense to store instants - we can return what we want, but since the type is DATE_TIME it seemed nice to return a type with helper functions for dates.